### ISSUE TITLE
Invalid path for TLS certificates but certificates accessible to envoy user within the container

### ISSUE THREAD

**Issue by chintan-mishra** [envoyproxy/envoy#13221]
*Title*: *TLS certificates provided by LetsEncrypt when used in envoy.yaml config throw error showing `invalid path`*

*Description*:
TLS certificates available to envoy user are not found while starting Envoy using Podman on Fedora 32. The container is started using the command below.
```
$ sudo podman run -p 80:8080 -p 443:8443 -p 3306:3306 \
    -v /etc/envoy/envoy.yaml:/etc/envoy/envoy.yaml:Z \
    -v /etc/letsencrypt/:/etc/letsencrypt/:Z \
    --network dev --name dev-envoy --security-opt label=type:envoy.process \
    -d envoyproxy/envoy:v1.15.0
```
The container cannot find the TLS certificates and returns the following error:

```
[2020-09-22 18:15:55.556][3][critical][main] [source/server/server.cc:101] error initializing configuration '/etc/envoy/envoy.yaml': Invalid path: /etc/letsencrypt/live/rebhu.com/cert.pem
[2020-09-22 18:15:55.557][3][info][main] [source/server/server.cc:704] exiting
Invalid path: /etc/letsencrypt/live/rebhu.com/cert.pem
```
These files are present in the container and accessible to the container's envoy user. [Here](https://matrix.to/#/!xsXlHPPkTHzgsTONtn:matrix.org/$160079698852868UtfcN:matrix.org?via=matrix.org&via=tout.im&via=darmstadt.social) is the IRC-Matrix.org conversation I had with the developers of Podman. The `envoy.yaml` config and the TLS certificates share the same permissions. However, the certificates are not accessible by the application but the configuration file is. This is how we can access the envoy user within the container and at least read the TLS certificate files.

Before running the following command, comment out the `listener_mysql` and `listener_https` sections in the configuration attached below. Otherwise, the container will return the same error.
```
$ sudo podman container rm dev-envoy
$ sudo podman run -p 80:8080 \
    -v /etc/envoy/envoy.yaml:/etc/envoy/envoy.yaml:Z \
    -v /etc/letsencrypt/:/etc/letsencrypt/:Z \
    --network dev --name dev-envoy --security-opt label=type:envoy.process \
    -d envoyproxy/envoy:v1.15.0
$ sudo podman exec -it --user envoy dev-envoy bash
envoy@<container-id>$ cat /etc/letsencrypt/live/rebhu.com/cert.pem
```

>Any extra documentation required to understand the issue.

<details>
  <summary>Here is a small introduction on how to run Envoy using Podman on Fedora 32</summary>

```shell
$ sudo dnf install podman dnsmasq udica
$ sudo podman pull envoyproxy/envoy:v1.15.0
$ sudo podman create -p 80:8080 -p 443:8443 -p 3306:3306 \
    -v /etc/envoy/envoy.yaml:Z -v /etc/letsencrypt/:/etc/letsencrypt:Z \
    --network dev --name my-envoy envoyproxy/envoy:v1.15.0
```
Using `run` in place of `create` returns the following error
```shell
# chown: changing ownership of '/dev/stdout': Permission denied
# chown: changing ownership of '/dev/stderr': Permission denied
```
We can avoid this by following these steps.
```shell
$ sudo podman inspect envoy > my-envoy.udica.json
$ sudo udica -j my-envoy.udica.json my-envoy
```
This will generate a SELinux `*.cil` policy file. Use this file to activate SELinux policy specific to the Envoy container. Before activating the policy you will have to handle some Access Vector Cache(AVC) issues. This can be done by editing the newly generated `*.cil` file. A reference file has been added below. You can monitor the AVC using the command below.
```shell
$ sudo cat /var/log/audit/audit.log | grep AVC
```
After editing the `*.cil` file, install the SELinux policy using the following command.
```shell
$ sudo semodule -i my-envoy.cil /usr/share/udica/templates/{base_container.cil,net_container.cil}
```
Now our SELinux policy is loaded and we can use it to start our container. Notice the difference in the executed command. `-p 443:8443` and `-p 3306:3306` have been removed as these are the reason for raising this issue. Before running the following command, comment out the `listener_mysql` and `listener_https` sections in the configuration attached below. Otherwise, the container will return the same error.
```shell
$ sudo podman run -p 80:8080 \
    -v /etc/envoy/envoy.yaml:/etc/envoy/envoy.yaml:Z \
    -v /etc/letsencrypt/:/etc/letsencrypt/:Z \
    --network dev --name dev-envoy --security-opt label=type:my-envoy.process -d \
    envoyproxy/envoy:v1.15.0
```  
</details>

<details>
<summary>`*.cil` file for starting Envoy</summary>

This is the `*.cil` file that will get generated by the Udica tool. However, Udica can only generate what it sees from the `podman inspect` command. And it doesn't know about the application specific error. This configuration has allowed HTTP(S), MySQL, and volume mount related details.
```
(block envoy
    (blockinherit container)
    (blockinherit restricted_net_container)
    (allow process process ( capability ( audit_write chown dac_override fowner fsetid kill mknod net_bind_service net_raw setfcap setgid setpcap setuid sys_chroot ))) 

    (allow process mysqld_port_t ( tcp_socket (  name_bind ))) 
    (allow process http_cache_port_t ( tcp_socket (  name_bind ))) 
    (allow process http_port_t ( tcp_socket (  name_bind ))) 
    (allow process etc_t ( dir ( add_name create getattr ioctl lock open read remove_name rmdir search setattr write ))) 
    (allow process etc_t ( file ( append create getattr ioctl lock map open read rename setattr unlink write ))) 
    (allow process etc_t ( sock_file ( append getattr open read write ))) 
    (allow process etc_t ( dir ( add_name create getattr ioctl lock open read remove_name rmdir search setattr write ))) 
    (allow process etc_t ( file ( append create getattr ioctl lock map open read rename setattr unlink write ))) 
    (allow process etc_t ( sock_file ( append getattr open read write )))
)
```
Before installing(`sudo semodule -i ....`) the SELinux policy, append the following lines in the file above.
```
    (allow process container_runtime_t ( fifo_file ( setattr )))
    (allow process unreserved_port_t (tcp_socket (name_bind)))
```
</details>

<details>
<summary>Here is Envoy's configuration that I am using</summary>

```YAML
admin:
  access_log_path: /tmp/admin_access.log
  address:
    socket_address:
      protocol: TCP
      address: 127.0.0.1
      port_value: 9901
static_resources:
  listeners:
  - name: listener_http
    address:
      socket_address:
        protocol: TCP
        address: 0.0.0.0
        port_value: 8080
    filter_chains:
    - filters:
      - name: envoy.filters.network.http_connection_manager
        typed_config:
          "@type": type.googleapis.com/envoy.config.filter.network.http_connection_manager.v2.HttpConnectionManager
          stat_prefix: ingress_http
          route_config:
            name: http_route
            virtual_hosts:
            - name: http_service
              domains: ["*"]
              routes:
              - match:
                  prefix: "/.well-known"
                route:
                  cluster: service_letsencrypt
              - match:
                  prefix: "/.well-known/"
                route:
                  cluster: service_letsencrypt
          http_filters:
          - name: envoy.filters.http.router
  - name: listenet_https
    address:
      socket_address:
        protocol: TCP
        address: 0.0.0.0
        port_value: 8443
    filter_chains:
    - filter_chain_match:
        server_names: ["rebhu.com"]
      transport_socket:
        name: envoy.transport_sockets.tls
        typed_config:
          "@type": type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.DownstreamTlsContext
          common_tls_context:
            tls_certificates:
            - certificate_chain:
                filename: "/etc/letsencrypt/live/rebhu.com/fullchain.pem"
              private_key:
                filename: "/etc/letsencrypt/live/rebhu.com/privkey.pem"
      filters:
      - name: envoy.filters.network.https_connection_manager
        typed_config:
          "@type": type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager
          stat_prefix: ingress_https
          route_config:
            virtual_hosts:
            - name: wp_https
              domains: "*"
              routes:
              - match:
                  prefix: "/"
                route:
                  cluster: service_letsencrypt
  - name: listener_mysql
    address:
      socket_address:
        protocol: TCP
        address: 0.0.0.0
        port_value: 3306
    filter_chains:
    - filter_chain_match:
        server_names: ["db.rebhu.com"]
      transport_socket:
        name: envoy.transport_sockets.tls
        typed_config:
          "@type": type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.DownstreamTlsContext
          common_tls_context:
            tls_certificates:
            - certificate_chain:
                filename: "/etc/letsencrypt/live/rebhu.com/cert.pem"
              private_key:
                filename: "/etc/letsencrypt/live/rebhu.com/privkey.pem"
      filters:
      - name: envoy.filter.network.mysql_proxy
        typed_config:
          "@type": "type.googleapis.com/envoy.extensions.filters.network.mysql_proxy.v3.MySQLProxy"
          stat_prefix: ingress_mysql
      - name: envoy.filters.network.tcp_proxy
        typed_config:
          "@type": type.googleapis.com/envoy.extensions.filters.network.tcp_proxy.v3.TcpProxy
          stat_prefix: ingress_tcp_mysql
          cluster: service_mysql

  clusters:
  - name: service_letsencrypt
    connect_timeout: 0.25s
    type: LOGICAL_DNS
    dns_lookup_family: V4_ONLY
    lb_policy: ROUND_ROBIN
    load_assignment:
      cluster_name: service_letsencrypt
      endpoints:
      - lb_endpoints:
        - endpoint:
            address:
              socket_address:
                address: dev-nginx
                port_value: 80
  - name: service_mysql
    connect_timeout: 0.25s
    type: LOGICAL_DNS
    dns_lookup_family: V4_ONLY
    lb_policy: ROUND_ROBIN
    load_assignment:
      cluster_name: service_mysql
      endpoints:
      - lb_endpoints:
        - endpoint:
            address:
              socket_address:
                address: dev-mysql
                port_value: 3306
```
</details>

**Comment by chintan-mishra** [comment]
Any suggestions to get this working on Fedora/CentOS with Podman?

**Comment by phlax** [comment]
@chintan-mishra can you check the permissions of the files inside the container.

im wondering if the problem is that the `envoy` user cannot read these files. iirc letsencrypt creates non-other-user permissions for certs, so i guess that is why

**Comment by phlax** [comment]
sorry, reading issue properly (and title) and realising that the envoy user *can* read these files, so not sure

just to check can you do

```
ls -alh /etc/letsencrypt/live/rebhu.com/
```

inside container as envoy user

**Comment by phlax** [comment]
im not so familiar with podman - but my understanding is that it will use entrypoints ~the same as docker

on that basis, you dont want to start the container using the `--user` flag - let the entrypoint change user

on docker, eg:

```
$ docker run --rm -it  envoyproxy/envoy:v1.15.0 whoami
envoy
```

**Comment by phlax** [comment]
not sure if that is related to your  problem, but would be good to just exclude as source of confusion at least

**Comment by chintan-mishra** [comment]
As mentioned earlier, I started the container after deleting the `listener_https` in the attached Envoy YAML config.

> im not so familiar with podman - but my understanding is that it will use entrypoints ~the same as docker
Yes. For most parts, Podman is a drop in replacement for Docker

Here is the output of `whoami` command
```bash
[fedora@ip-172-0-1-7 ~]$ sudo podman run --rm -it envoyproxy/envoy:v1.16.0 whoami
envoy
```
> just to check can you do
> `ls -alh /etc/letsencrypt/live/rebhu.com/`
> inside container as envoy user

```bash
[fedora@ip-172-0-1-7 ~]$ sudo podman exec -it --user envoy test-envoy bash
envoy@3ae******363:/$ ls -alh /etc/letsencrypt/live/rebhu.com/ 
total 12K
drwx------. 2 root root 4.0K Sep 22 10:24 .
drwxrwx---. 3 root root 4.0K Sep 22 10:24 ..
-rw-r--r--. 1 root root  692 Sep 22 10:24 README
lrwxrwxrwx. 1 root root   36 Sep 22 10:24 cert.pem -> ../../archive/rebhu.com/cert1.pem
lrwxrwxrwx. 1 root root   37 Sep 22 10:24 chain.pem -> ../../archive/rebhu.com/chain1.pem
lrwxrwxrwx. 1 root root   41 Sep 22 10:24 fullchain.pem -> ../../archive/rebhu.com/fullchain1.pem
lrwxrwxrwx. 1 root root   39 Sep 22 10:24 privkey.pem -> ../../archive/rebhu.com/privkey1.pem
```

**Comment by phlax** [comment]
so its showing symlinks - can you do `ls -lh` to the actual files

**Comment by chintan-mishra** [comment]
Here is the output you requested
```
envoy@3ae******363:/$ ls -lh /etc/letsencrypt/archive/rebhu.com/
total 16K
-rw-r--r--. 1 root root 2.1K Sep 22 10:24 cert1.pem
-rw-r--r--. 1 root root 1.7K Sep 22 10:24 chain1.pem
-rw-r--r--. 1 root root 3.7K Sep 22 10:24 fullchain1.pem
-rw-------. 1 root root 1.7K Sep 22 10:24 privkey1.pem
```

**Comment by phlax** [comment]
hmm. not sure - it seems like `cert.pem` is readable so that doesnt seem to be the trouble

**Comment by phlax** [comment]
its not ideal but how about trying with the non-symlinked path and seeing if that works ?

**Comment by chintan-mishra** [comment]
```
[2020-11-17 12:59:10.970][3][critical][main] [source/server/server.cc:102] error initializing configuration '/etc/envoy/envoy.yaml': Invalid path: /etc/letsencrypt/archive/rebhu.com/cert1.pem
```

I got called into some work yesterday. Here is the output after switching from symlinks to the source files

**Comment by phlax** [comment]
im not sure. it might  take me a day or two but i can set up podman locally and see if i can repro the problem

**Comment by chintan-mishra** [comment]
Do let me know any issues you face while working with Podman. I have raised this issue with the Podman team on IRC.

Podman has better tooling for SELinux. The tool `udica` only generates SELinux policies. It doesn't cover AppArmor. I believe AppArmor ships by default in Debian 10 and its derivatives.

**Comment by chintan-mishra** [comment]
@phlax any luck with this issue?

**Comment by chintan-mishra** [comment]
Since Podman 2.2, I cannot access TLS certificate files as `envoy` user within the container. Podman IRC and mailing list are quiet for some time. I will try again in 2021. This release is bug-for-bug compatible with Docker.

Some problems mentioned in the issues have crept up in the v1.14.6 as well. I have yet to determine if these are from the upgrades in Envoy or Podman.

I will inspect further and share my findings.

Happy 2021!!

**Comment by chintan-mishra** [comment]
There is a hack to get the `envoy` user to read the certficates.(Hint: Use `chown` and customize Envoy container)

However, after scanning a lot of posts related to Certbot and LetsEncrypt on relevant forums, this hack is not recommended. Permissions related to these files are controlled by Certbot application and the `root` user of the system.

This makes me wonder, is anyone else facing issues with mounting TLS certificates provided by LetsEncrypt with Docker?

**Comment by lamualfa-client** [comment]
Same issue with me. I use Envoy 1.17

**Comment by chintan-mishra** [comment]
@lamualfa-client you might find this helpful.

@phlax suggested three possible fix for this on #14877 

1. Change the permission of the TLS certs or any other files for that matter before mounting them on the envoy container,
2. Create a custom container and within the container change the file permissions allowing the `envoy` user to work with them
3. Add `-e ENVOY_UID=0` while starting EnvoyProxy

I have tested 1, and 3 and these two work for me.

**Comment by arbfay** [comment]
Had the same issue. `ENVOY_UID=0` worked for me. Thanks @chintan-mishra
